Add full node documentation to RC1 (#3567)
celo-org/celo-monorepo: packages/docs/getting-started/running-a-full-node-in-rc1.md

# Running a Full Node

This section explains how to get a full node running on the [Release Candidate 1 Network](rc1.md), using a Docker image that was built for this purpose.

Full nodes play a special purpose in the Celo ecosystem, acting as a bridge between the mobile wallets \(running as light clients\) and the validator nodes. To make sure that full nodes are rewarded for this service, the Celo protocol includes [full node incentives](../celo-codebase/protocol/transactions/full-node-incentives.md).

> When a light client sends transactions, they will include specify a gateway fee to be paid to the node that gossips transactions to the other full nodes and validators.

Reviewer: note to double check this -- I don't think this is enabled already in RC1?

Author: @asaj do we know when full node incentives will be enabled?

Reviewer 2: This is not yet enabled because the mechanism for negotiating a gateway fee is not yet built. Currently, light clients are set to send a gateway fee of 0 and full nodes are set to accept a minimum gatewayfee of 0

Author: Added a warning to explain this.

> For this reason, despite the fact that Celo uses a proof-of-stake protocol, users can earn cryptocurrency without first having to own any, simply by running a full node.

Reviewer: ```suggestion
For this reason, despite the fact that Celo uses a proof-of-stake consensus protocol, users can earn cryptocurrency without first having to own any, simply by running a full node.
```

## Prerequisites

- **You have Docker installed.** If you don’t have it already, follow the instructions here: [Get Started with Docker](https://www.docker.com/get-started). It will involve creating or signing in with a Docker account, downloading a desktop app, and then launching the app to be able to use the Docker CLI. If you are running on a Linux server, follow the instructions for your distro [here](https://docs.docker.com/install/#server). You may be required to run Docker with `sudo` depending on your installation environment.

{% hint style="info" %}
Code you'll see on this page is bash commands and their output.

When you see text in angle brackets &lt;&gt;, replace them and the text inside with your own value of what it refers to. Don't include the &lt;&gt; in the command.
{% endhint %}

## Celo Networks

> First we are going to setup the environment variables required for `Baklava` network. Run:

Reviewer: ```suggestion
First we are going to setup the environment variables required for the `RC1` network. Run:
```

```bash
export CELO_IMAGE=us.gcr.io/celo-testnet/celo-node:rc1
export NETWORK_ID=42220
```

## Pull the Celo Docker image

We're going to use a Docker image containing the Celo node software in this tutorial.

If you are re-running these instructions, the Celo Docker image may have been updated, and it's important to get the latest version.

```bash
docker pull $CELO_IMAGE
```

## Set up a data directory

First, create the directory that will store your node's configuration and its copy of the blockchain. This directory can be named anything you'd like, but here's a default you can use. The commands below create a directory and then navigate into it. The rest of the steps assume you are running the commands from inside this directory.

```bash
mkdir celo-data-dir
cd celo-data-dir
```

## Create an account and get its address

In this step, you'll create an account on the network. If you've already done this and have an account address, you can skip this and move on to configuring your node.

Run the command to create a new account:

```bash
docker run -v $PWD:/root/.celo --rm -it $CELO_IMAGE account new
```

> It will prompt you for a passphrase, ask you to confirm it, and then will output your account address: `Address: {<YOUR-ACCOUNT-ADDRESS>}`

Reviewer: think this format may have changed.. note to double check what the output looks like now

Save this address to an environment variables, so that you can reference it below (don't include the braces):

```bash
export CELO_ACCOUNT_ADDRESS=<YOUR-ACCOUNT-ADDRESS>
```

{% hint style="info" %}
This environment variable will only persist while you have this terminal window open. If you want this environment variable to be available in the future, you can add it to your `~/.bash_profile`
{% endhint %}

## Configure the node

> The genesis block is the first block in the chain, and is specific to each network. This command gets the `genesis.json` file for baklava and uses it to initialize your nodes' data directory.

Reviewer: ```suggestion
The genesis block is the first block in the chain, and is specific to each network. This command gets the `genesis.json` file for RC1 and uses it to initialize your nodes' data directory.
```

```bash

> docker run -v $PWD:/root/.celo --rm $CELO_IMAGE init /celo/genesis.json

Reviewer: do we need to include `-it` flags here?

Reviewer 3: Done

```

In order to allow the node to sync with the network, get the enode URLs of the bootnodes:

```bash
export BOOTNODE_ENODES=$(docker run --rm --entrypoint cat $CELO_IMAGE /celo/bootnodes)
```

## Start the node

This command specifies the settings needed to run the node, and gets it started.

```bash

> docker run --name celo-fullnode -d --restart unless-stopped -p 127.0.0.1:8545:8545 -p 127.0.0.1:8546:8546 -p 30303:30303 -p 30303:30303/udp -v $PWD:/root/.celo $CELO_IMAGE --verbosity 3 --networkid $NETWORK_ID --syncmode full --rpc --rpcaddr 0.0.0.0 --rpcapi eth,net,web3,debug,admin,personal --light.serve 90 --light.maxpeers 1000 --maxpeers 1100 --etherbase $CELO_ACCOUNT_ADDRESS --bootnodes $BOOTNODE_ENODES --nousb

Reviewer: do we need to include `--gcmode=archive`?

Reviewer: also, would suggest adding a hint that allows you to run with Ledger :)

Reviewer 2: You shouldn't need to run an archive node

```

> You'll start seeing some output. After a few minutes, you should see lines that look like this. This means your node has synced with the network and is receiving blocks.

Reviewer: ```suggestion
You'll start seeing some output. After a few minutes, you should see lines that look like this. This means your node has started syncing with the network and is receiving blocks.
```

```text
INFO [07-16|14:04:24.924] Imported new chain segment               blocks=139  txs=319 mgas=61.987 elapsed=8.085s mgasps=7.666 number=406  hash=9acf16…4fddc8 age=6h58m44s cache=1.51mB
INFO [07-16|14:04:32.928] Imported new chain segment               blocks=303  txs=179 mgas=21.837 elapsed=8.004s mgasps=2.728 number=709  hash=8de06a…77bb92 age=6h33m37s cache=1.77mB
INFO [07-16|14:04:40.918] Imported new chain segment               blocks=411  txs=0   mgas=0.000  elapsed=8.023s mgasps=0.000 number=1120 hash=3db22a…9fa95a age=5h59m30s cache=1.92mB
INFO [07-16|14:04:48.941] Imported new chain segment               blocks=335  txs=0   mgas=0.000  elapsed=8.023s mgasps=0.000 number=1455 hash=7eb3f8…32ebf0 age=5h31m43s cache=2.09mB
INFO [07-16|14:04:56.944] Imported new chain segment               blocks=472  txs=0   mgas=0.000  elapsed=8.003s mgasps=0.000 number=1927 hash=4f1010…1414c1 age=4h52m31s cache=2.34mB
```

> You will have fully synced with the network once you have pulled the latest block number, which you can lookup by visiting at the [Baklava Testnet Stats](https://baklava-celostats.celo-testnet.org/) or [Baklava Network Explorer]](https://baklava-blockscout.celo-testnet.org/) pages.

Reviewer: update with RC1 links?
